### PR TITLE
move 3rd party types to dev dependencies in CLI NPM package

### LINKED ISSUE #3371: Move all @types CLI dependencies from prod to dev list
## Problem

When we ship Cypress NPM package we include several `@types` prod dependencies because we want to reference them in our types `index.d.ts` file. 

in `cli/package.json`

```json
{
  "dependencies": {
    "@types/blob-util": "1.3.3",
    "@types/bluebird": "3.5.18",
    "@types/chai": "4.0.8",
    "@types/chai-jquery": "1.1.35",
    "@types/jquery": "3.3.6",
    "@types/lodash": "4.14.87",
    "@types/minimatch": "3.0.3",
    "@types/mocha": "2.2.44",
    "@types/sinon": "7.0.0",
    "@types/sinon-chai": "3.2.2"
  }
}
```

and in our `index.d.ts` file we reference these dependencies (and a few local files)

```ts
/// <reference path="./blob-util.d.ts" />
/// <reference path="./bluebird.d.ts" />
/// <reference path="./minimatch.d.ts" />
/// <reference path="./moment.d.ts" />

/// <reference types="chai" />
/// <reference types="chai-jquery" />
/// <reference types="jquery" />
/// <reference types="lodash" />
/// <reference types="mocha" />
/// <reference types="moment" />
/// <reference types="sinon" />
/// <reference types="sinon-chai" />
```

We need these type files because they provide types and documentation for things like Lodash `Cypress._` object, but there are a couple of problems

1. We have duplicates: for example, we depend on `@types/blob-util` but also have a local file `blob-util.d.ts`
2. type dependencies in _prod_ can conflict with user's type dependencies, which seems to the issue in #1227
3. usually `@types` is considered a _dev_ dependency, should not be _prod_

## Proposal

We can move all necessary `@types` packages from `dev` to `prod` dependencies. During building we can copy any necessary files and include them with our CLI code, just like we do for `types/blob-util.d.ts`

Our distributed `index.d.ts` thus will only refer to local paths to bring it all together, and all conflicts with the user's `@types` would be avoided (fingers crossed)

### PR DESCRIPTION
**Pull request by bahmutov**
- closes #3371
- [x] move `@types/lodash`
- [x] move `@types/blob-util`
- [x] move `@types/bluebird`
- [x] move `@types/minimatch`
- [x] move `@types/jquery`
- [x] move `@types/chai`
- [x] move `@types/chai-jquery`
- [x] move `@types/mocha`
- [x] move `@types/sinon`
- [x] move `@types/sinon-chai`

### CODE REVIEW

**Comment by bahmutov** [comment]
@tgriesser can you take a look at the typescript changes - read the issue for motivation, trying to prevent type clashes between our bundled types and users' (like Lodash)

@NicholasBoll if you can take a quick look, we would appreciate it. Chai and jQuery are weird types since they are global and I had to delete `node_modules/@types/chai/index.d.ts` and `node_modules/@types/jquery/index.d.ts` to get `dtslint` to pass

**Comment by tgriesser** [comment]
This makes sense that we'd want to avoid including types that would conflict with users' installed types. 

I'm wondering if there's any tooling that exists already that will solve this problem more generally for us, by effectively bundling them inline when they're imported so we have one (or two) big d.ts files which handles the namespacing, etc.

https://github.com/timocov/dts-bundle-generator
https://github.com/SitePen/dts-generator
https://github.com/TypeStrong/dts-bundle

I might play around with this because I've been wanting to solve this problem for knex as well, as it currently has `@types/bluebird` as [a dependency](https://github.com/tgriesser/knex/blob/master/package.json#L11) for similar reasons as Cypress.

But generally I think the approach here makes sense, without getting it running i'm not sure if there's any issues with it.

**Comment by bahmutov** [comment]
@tgriesser I have looked at the TS bundle generators, they would need to tested well, since all require extensive configuration, and I have never used them. For now I think copying dependencies would do, and I will open a separate issue to find and use TS bundler.

**Comment by NicholasBoll** [comment]
At first I thought this was a terrible idea, but now I like it. The problem is around global definitions (namespaces) that collide. I looked into jQuery type definitions and chai type definitions. It looks like jQuery breaks up namespacing into different files, but chai looks like its namespace is in the same file. Does deleting it cause type errors where the chai type is not found?

Do you want me to try out these changes to verify it will work externally?

**Comment by bahmutov** [comment]
ughh, typescript@next does not like this change at all

```
Installing to /root/.dts/typescript-installs/next...
Installed!
Error: /root/cypress/cli/types/cy-chai.d.ts:3:22
ERROR: 3:22      expect  TypeScript@next compile error: 
File '/root/cypress/cli/types/chai/index.d.ts' not found.
ERROR: 9:28      expect  TypeScript@next compile error: 
Cannot find name 'chai'.
ERROR: 10:30     expect  TypeScript@next compile error: 
Cannot find name 'chai'.
ERROR: 11:30     expect  TypeScript@next compile error: 
Cannot find name 'chai'.

/root/cypress/cli/types/index.d.ts:19:22
ERROR: 19:22     expect  TypeScript@next compile error: 
File '/root/cypress/cli/types/jquery/index.d.ts' not found.
ERROR: 82:8      expect  TypeScript@next compile error: 
Cannot find name 'JQueryStatic'.
ERROR: 274:25    expect  TypeScript@next compile error: 
Cannot find name 'JQuery'.
```

and so on in https://circleci.com/gh/cypress-io/cypress/75061

**Comment by bahmutov** [comment]
Testing build NPM package
```
$ npm i https://cdn.cypress.io/beta/npm/3.1.6/circle-move-ts-types-to-dev-dependencies-3371-02202412002cfd0541cae6096fc614863750795d-76512/cypress.tgz --ignore-scripts
```

against kitchensink still has a problem
```
$ npm run types

> cypress-example-kitchensink@0.0.0-development types /Users/gleb/git/cypress-example-kitchensink
> tsc --noEmit

node_modules/cypress/types/cy-minimatch.d.ts:5:28 - error TS7016: Could not find a declaration file for module 'minimatch'. '/Users/gleb/git/cypress-example-kitchensink/node_modules/minimatch/minimatch.js' implicitly has an 'any' type.
  Try `npm install @types/minimatch` if it exists or add a new declaration (.d.ts) file containing `declare module 'minimatch';`

5 import * as mimimatch from 'minimatch'
                             ~~~~~~~~~~~

node_modules/cypress/types/sinon-chai/index.d.ts:12:24 - error TS2307: Cannot find module 'sinon'.

12 import * as Sinon from 'sinon';
```

**Comment by bahmutov** [comment]
I think the TS problems have been solved, all 3rd party deps use relative paths

File: circle.yml [hunk 1]
```diff
@@ -754,6 +763,7 @@ linux-workflow: &linux-workflow
           branches:
             only:
               - develop
+              - move-ts-types-to-dev-dependencies-3371
```
Comment (Saibamen) [inline circle.yml]: This should be deleted

**Comment by avahe-kellenberger** [comment]
Is this supposed to solve cypress installed `@types/` packages? It's still installing `@types/chai-jquery` for me, currently.

I also saw [post-install.js](https://github.com/cypress-io/cypress/blob/83d35d37255ee5d5b5c01b27bcc621245a324a71/cli/scripts/post-install.js) which seemed like it would take care of it, but `yarn install cypress -D` is still installing `chai-jquery` under types.

**Pull request merged**